File: agent/src/api/sse.ts
import type { ServerResponse } from 'node:http';

import { runFailed, type AgentEvent } from '../runtime/events.js';

export function encodeSse(event: AgentEvent): string {
  return `event: ${event.type}\ndata: ${JSON.stringify({
    runId: event.runId,
    payload: event.payload,
  })}\n\n`;
}

/** 把 Runtime 的异步事件写入长连接，并定时发送 SSE 注释作为心跳。 */
export async function writeSseStream(
  response: ServerResponse,
  events: AsyncIterable<AgentEvent>,
  runId: string,
  heartbeatMs: number,
): Promise<void> {
  const heartbeat = setInterval(() => {
    if (!response.destroyed) response.write(': keepalive\n\n');
  }, heartbeatMs);

  try {
    for await (const event of events) {
      if (response.destroyed) break;
      response.write(encodeSse(event));
    }
  } catch (error) {
    console.error('Agent SSE stream failed', { runId, error });
    if (!response.destroyed) {
      response.write(
        encodeSse(runFailed(runId, 'AGENT_SERVICE_ERROR', 'Agent 服务执行失败', true)),
      );
    }
  } finally {
    clearInterval(heartbeat);
    if (!response.destroyed) response.end();
  }
}

File: agent/eterion_agent/api/sse.py
"""SSE transport helpers kept separate from Agent execution."""

from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator
from contextlib import suppress
import json
import logging

from eterion_agent.runtime import AgentEvent, failure


logger = logging.getLogger(__name__)


async def stream_with_heartbeats(
    events: AsyncIterator[AgentEvent],
    heartbeat_seconds: float,
    run_id: str,
) -> AsyncIterator[str]:
    iterator = events.__aiter__()
    pending: asyncio.Task[AgentEvent] | None = asyncio.create_task(anext(iterator))
    try:
        while pending is not None:
            done, _ = await asyncio.wait({pending}, timeout=heartbeat_seconds)
            if not done:
                yield ": keepalive\n\n"
                continue
            try:
                event = pending.result()
            except StopAsyncIteration:
                pending = None
                break
            yield encode_sse(event)
            pending = asyncio.create_task(anext(iterator))
    except asyncio.CancelledError:
        raise
    except Exception:
        logger.exception("Agent SSE stream failed")
        yield encode_sse(
            failure(run_id, "AGENT_SERVICE_ERROR", "Agent 服务执行失败", True)
        )
    finally:
        if pending is not None and not pending.done():
            pending.cancel()
            with suppress(asyncio.CancelledError):
                await pending
        with suppress(Exception):
            await iterator.aclose()


def encode_sse(event: AgentEvent) -> str:
    data = {"runId": event.run_id, "payload": event.payload}
    payload = json.dumps(data, ensure_ascii=False, separators=(",", ":"))
    return f"event: {event.type}\ndata: {payload}\n\n"
